packages/flutter/lib/src/widgets/focus_scope.dart

```diff
@@ -1,145 +1,422 @@
+    this.focusNode,
     this.autofocus = false,
-    this.child,
-  }) : assert(node != null),
-       assert(autofocus != null),
-       super(key: key);
-
-  /// Controls whether this scope is currently active.
-  final FocusScopeNode node;
+    this.onFocusChange,
+    this.onKey,
+    this.debugLabel,
+  })  : assert(child != null),
+        assert(autofocus != null),
+        super(key: key);
 
-  /// Whether this scope should attempt to become active when first added to
-  /// the tree.
-  final bool autofocus;
+  /// A debug label for this widget.
+  ///
+  /// Not used for anything except to be printed in the diagnostic output from
+  /// [toString] or [toStringDeep].
+  ///
+  /// To see the entire tree, call [debugDescribeFocusTree].
+  ///
+  /// Defaults to null.
+  final String debugLabel;
 
-  /// The widget below this widget in the tree.
+  /// The child widget of this [Focus].
   ///
   /// {@macro flutter.widgets.child}
   final Widget child;
 
-  /// Returns the [node] of the [FocusScope] that most tightly encloses the
-  /// given [BuildContext].
+  /// Handler for keys pressed when this object or one of its children has
+  /// focus.
   ///
-  /// The [context] argument must not be null.
-  static FocusScopeNode of(BuildContext context) {
-    assert(context != null);
-    final _FocusScopeMarker scope = context.inheritFromWidgetOfExactType(_FocusScopeMarker);
-    return scope?.node ?? context.owner.focusManager.rootScope;
-  }
+  /// Key events are first given to the leaf nodes, and if their [onKey] methods
+  /// return false, then they are given to each ancestor node up the focus
+  /// hierarchy in turn. If an event reaches the root of the hierarchy, it is
+  /// discarded.
+  ///
+  /// This is not the way to get text input in the manner of a text field: it
+  /// leaves out support for input method editors, and doesn't support soft
+  /// keyboards in general. For text input, consider [TextField],
+  /// [EditableText], or [CupertinoTextField] instead, which do support these
+  /// things.
+  final FocusOnKeyCallback onKey;
+
+  /// Handler called when the focus changes.
+  ///
+  /// Called with true if this node gains focus, and false if it loses
+  /// focus.
+  final ValueChanged<bool> onFocusChange;
+
+  /// True if this widget will be selected as the initial focus when no other
+  /// node in its scope is currently focused.
+  final bool autofocus;
 
-  /// A list of the [FocusScopeNode]s for each [FocusScope] ancestor of
-  /// the given [BuildContext]. The first element of the list is the
-  /// nearest ancestor's [FocusScopeNode].
+  /// An optional focus node to use as the focus node for this [Focus] widget.
   ///
-  /// The returned list does not include the [FocusManager]'s `rootScope`.
-  /// Only the [FocusScopeNode]s that belong to [FocusScope] widgets are
-  /// returned.
+  /// If one is not supplied, then one will be allocated and managed by this
+  /// widget.
+  final FocusNode focusNode;
```

Comment:
Is there a good reason to ever provide one manually? (I thought the whole point of the existence of this widget was to not manually manage focus).

Reply:
Yes.  There are cases where you want to have an ancestor own the node, and pass it in to this widget to host, since you might (for instance, in `CupertinoTabScaffold`) want to have the owning widget tell its children to take focus, and the easiest way to do that is to own the node, and give it to them to host.